### DISCUSSION TITLE
filer.backup: More features + filer.restore

### DISCUSSION THREAD

**Discussion by KevinBLT** [seaweedfs/seaweedfs#3469] [Ideas]
Hello all!

I've read through all documentations and issues and cannot find more informations about some topics I am going to share. Generally they are some ideas I've had and would love to see in SeaweedFS! I think they might be easy for Chris of course but any experienced go developer as well - which is not me, otherwise I would have loved to create a pull request. I thought about implementing in another language but I thought these features would be better for a broader audience of users.
____________________
Imagine a `weed filer.backup` running with `is_incremental=true`.

#### Feature 1
From a filesystem level view I would at least like to know, that I file was deleted on a specific day.
I would think such an implementation would be pretty ease. Instead of `if (file.deleted) return`, one would create a prefixed symlink to the file with `deleted.{filename}`.

As an example:
```bash
kb@host_dev1:/volume1/docker/filer.backup/2022-08-20$ ls -l
total 32236
-rw-r----- 1 root root 28055812 Aug 20 20:05 file_a.txt
-rw-r----- 1 root root        9 Aug 20 21:01 deleted.file_d.log -> file_d.txt
-rw-r----- 1 root root  4574585 Aug 20 18:59 file_b.txt
-rw-r----- 1 root root    17271 Aug 20 20:23 file_d.txt
-rw-r----- 1 root root   177502 Aug 20 20:04 file_e.txt
-rw-r----- 1 root root   168872 Aug 20 19:56 file_f.txt
```

This would have the same benefits as currently but would give more insights about the past.

#### Feature 2

Enable a folder `current` on the level of the `YYYY-MM-DD` folders.
This folder would hold the actual state as in "the current filer" state.
To make this efficiently , `filer.backup` would only keep symlinks in a realtime state.

See this example to understand that better:
```bash
kb@host_dev1:/volume1/docker/filer.backup/current$ ls -l
-rw-r----- 1 root root 1024 Aug 20 20:05 file_a.txt -> ../2022-08-20/file_a.txt
-rw-r----- 1 root root 1024 Aug 20 20:05 file_b.txt -> ../2022-08-19/file_b.txt
-rw-r----- 1 root root 1024 Aug 20 20:05 file_c.txt -> ../2022-08-12/file_c.txt
-rw-r----- 1 root root 1024 Aug 20 20:05 file_d.txt -> ../2022-06-14/file_d.txt
-rw-r----- 1 root root 1024 Aug 20 20:05 file_e.txt -> ../2022-08-20/file_e.txt
```

When a file is deleted, the Symlink is deleted. When a file is updated, the symlink points to a newer folder.

In addition I would like a command like `weed filer.restore -day 2022-08-12`. This would create a folder like the `current` above maybe `state.2022-08-12` with all the symlinks.

#### Feature 3

This request is pretty short, I would like `-meta` parameter for `filer.backup` as I think it's often used together with `filer.meta.backup` and from my understanding both subscribe to changes etc.

### filer.restore

Together with the features above I would love a `weed filer.restore` reading from `replication.toml` where the sink was and restore a state on another filer from the folder `current` as mentioned above or a state folder (also mentioned above). Together with `-meta`, the `filer.restore` should be able to upload the files to a fresh restored file + do that with the backed up meta so collections etc. will also match again.

Please tell me what you think. I think these features shouldn't be so hard to implement and would help me and potentially many users to feel save about the backups and restore variants. 
Before I do that only for me in nodejs (for example) it would definitely fit perfectly into SeaweedFS itself.

**Comment by KevinBLT** [comment]
@chrislusf @kmlebedev What do you guys think? Or could you guide me to where these functions actually are, I have the feeling that this wouldn't be a lot of work, but I never used go in any way.

**Reply by chrislusf** [reply]
I feel these features are not so easy. Current logic is mainly in `replication/replicator.go`. Some refactoring is needed to support more custom backup behavior.

**Comment by kmlebedev** [comment]
@KevinBLT  It seems that points 1 and 2 are feasible if you send a PR

**Reply by KevinBLT** [reply]
Will try to do so, but will take some time since I have no knowledge of golang so far.
